File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Deque<Integer> stack = new ArrayDeque<>();
        int[] left = new int[heights.length];
        int[] right = new int[heights.length];

        for (int i = 0; i < heights.length; i++) {
            left[i] = -1;
            while (!stack.isEmpty() && heights[stack.peek()] >= heights[i]) {
                stack.pop();
            }
            if (!stack.isEmpty()) {
                left[i] = stack.peek();
            }
            stack.push(i);
        }
        
        stack.clear();
        for (int i = heights.length - 1; i >= 0; i--) {
            right[i] = heights.length;
            while(!stack.isEmpty() && heights[stack.peek()] >= heights[i]) {
                stack.pop();
            }
            if (!stack.isEmpty()) {
                right[i] = stack.peek();
            }
            stack.push(i);
        }

    int max = 0;
    for (int i =0; i < heights.length; i++){
        left[i] += 1;
        right[i] -= 1;
        max = Math.max(max, heights[i] * (right[i] - left[i] + 1));
    }
    return max;

    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int length = heights.size() - 1;
        int left = 0;
        int right = length;
        int max = 0;

        while (left != right) {
            int curr = std::min(heights[left], heights[right]) * length;
            if(curr > max) {
                max = curr;
            } else {
                if (heights[left] >= heights[right]) {
                    right--;
                } else {
                    left++;
                }
                length--;
            }
        }

        return max;
    }
};
